File: src/bumperbot_localization/bumperbot_localization/kalman_filter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu

class KalmanFilter(Node):

    def __init__(self):
        super().__init__("kalman_filter")
        self.odom_sub = self.create_subscription(Odometry, "bumperbot_controller/odom_noisy", self.odomCallback, 10)
        self.imu_sub = self.create_subscription(Imu, "imu/out", self.imuCallback, 10)
        self.odom_pub = self.create_publisher(Odometry, "bumperbot_controller/odom_kalman", 10)
        
        self.mean = 0.0
        self.variance = 1000.0
        self.motion_variance = 4.0
        self.measurement_variance = 0.5
        self.imu_angular_z = 0.0
        self.is_first_odom = True
        self.last_angular_z = 0.0
        self.motion = 0.0
        self.kalman_odom = Odometry()

    def odomCallback(self, odom):
        self.kalman_odom = odom

        if self.is_first_odom:
            self.last_angular_z = odom.twist.twist.angular.z
            self.is_first_odom = False
            self.mean = odom.twist.twist.angular.z
            return
        
        self.motion = odom.twist.twist.angular.z - self.last_angular_z
        self.statePrediction()
        self.measurementUpdate()
        self.last_angular_z = odom.twist.twist.angular.z
        self.kalman_odom.twist.twist.angular.z = self.mean
        self.odom_pub.publish(self.kalman_odom)

    def imuCallback(self, imu):
        self.imu_angular_z = imu.angular_velocity.z

    def measurementUpdate(self):
        self.mean = (self.measurement_variance * self.mean + self.variance * self.imu_angular_z) \
                   / (self.variance + self.measurement_variance)
                     
        self.variance = (self.variance * self.measurement_variance) \
                       / (self.variance + self.measurement_variance)

    def statePrediction(self):
        self.mean = self.mean + self.motion
        self.variance = self.variance + self.motion_variance

def main():
    rclpy.init()
    kalman_filter = KalmanFilter()
    rclpy.spin(kalman_filter)
    kalman_filter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_localization/src/kalman_filter.cpp
#include "bumperbot_localization/kalman_filter.hpp"

using std::placeholders::_1;

KalmanFilter::KalmanFilter(const std::string& name) 
                    : Node(name),
                      mean(0.0),
                      variance(1000.0),
                      motion_variance(4.0),
                      measurement_variance(0.5),
                      motion(0.0),
                      is_first_odom(true),
                      last_angular_z(0.0),
                      imu_angular_z(0.0)
{
  odom_sub = create_subscription<nav_msgs::msg::Odometry>("bumperbot_controller/odom_noisy", 10, std::bind(&KalmanFilter::odomCallback, this, _1));
  imu_sub = create_subscription<sensor_msgs::msg::Imu>("imu/out", 1000, std::bind(&KalmanFilter::imuCallback, this, _1));
  odom_pub = create_publisher<nav_msgs::msg::Odometry>("bumperbot_controller/odom_kalman", 10);
}

void KalmanFilter::odomCallback(const nav_msgs::msg::Odometry &odom)
{
  kalman_odom = odom;

  if(is_first_odom)
  {
      last_angular_z = odom.twist.twist.angular.z;
      mean = odom.twist.twist.angular.z;
      is_first_odom = false;
      return;
  }

    motion = odom.twist.twist.angular.z - last_angular_z;

    statePrediction();
    measurementUpdate();

    last_angular_z = odom.twist.twist.angular.z;

    kalman_odom.twist.twist.angular.z = mean;
    odom_pub->publish(kalman_odom);
}

void KalmanFilter::imuCallback(const sensor_msgs::msg::Imu &imu)
{
    imu_angular_z = imu.angular_velocity.z;
}

void KalmanFilter::measurementUpdate()
{
    mean = (measurement_variance * mean + variance * imu_angular_z)
          / (variance + measurement_variance);

    variance = (variance * measurement_variance) 
              / (variance + measurement_variance);
}

void KalmanFilter::statePrediction()
{
    mean = mean + motion;
    variance = variance + motion_variance;
}

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<KalmanFilter>("kalman_filter");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
